### ISSUE TITLE
Misleading "Fatal: unable to open config file" with s3 backend and no AWS* keys defined

### ISSUE THREAD

**Issue by paulschreiber** [restic/restic#1671]
## Output of `restic version`
```
restic 0.8.3
compiled with go1.10 on linux/amd64
```

## How did you run restic exactly?
```
restic -r s3:s3.amazonaws.com/backup.shrub.ca check
```

## What backend/server/service did you use to store the repository?
Amazon S3

## Expected behavior
Some sort of error, indicating that `AWS_ACCESS_KEY_ID`/`AWS_SECRET_ACCESS_KEY` is not defined.

## Actual behavior
```
Fatal: unable to open config file: Stat: Access Denied.
Is there a repository at the following location?
s3:s3.amazonaws.com/backup.shrub.ca
```

## Do you have an idea how to solve the issue?
If you define the expected variables, it runs:
```
# export AWS_ACCESS_KEY_ID=xxxx
# export AWS_SECRET_ACCESS_KEY=xxxx
# export RESTIC_PASSWORD=xxxx
# restic -r s3:s3.amazonaws.com/backup.shrub.ca check
```

**Comment by fd0** [comment]
Thanks for the idea, unfortunately it's not so easy: We support a few authentication methods for S3, see https://github.com/restic/restic/blob/0279fd72123adc52686156332ed7c00a3a07d6e9/internal/backend/s3/s3.go#L43-L64

And the error message already contains what's wrong:
```
Fatal: unable to open config file: Stat: Access Denied.
```

Do you have an idea how to solve this better?

**Comment by paulschreiber** [comment]
There's a difference between:
• user supplied credentials which were invalid
• user didn't supply any credentials

You're showing the same error in both cases. Further, the error is (somewhat) understandable to programmers, but not to end users.

Please indicate that no credentials were supplied and don't attempt to connect in this case.

**Comment by fd0** [comment]
Ah, thanks for the explanation.

**Comment by fd0** [comment]
I agree in general that "no credentials available" should be a different error than "wrong credentials". We need to consider more cases though, "user did not supply credentials" does not mean anything if the credentials can be acquired via IAM.

**Comment by harshavardhana** [comment]
> @harshavardhana is there a way to check if minio-go could not acquire any credentials at all?

Meaning you want to send an error when no credentials are present? @fd0 - That is not necessary as well it depends the bucket policy too so credentials might not have been specified. 

So the error is correct since the restic was used on a bucket where credentials are necessary. So "access denied" is what you will get for all calls.

**Comment by paulschreiber** [comment]
* I suppose it's technically possible that you'd allow world write to a bucket _and_ store your backups there. It just seems … unlikely.
* GCC 8 dramatically improved its error handling. A good summary of those improvements:
https://developers.redhat.com/blog/2018/03/15/gcc-8-usability-improvements/

---

This error message is confusing, even for sophisticated users. restic appears to be simply printing the underlying (programmer) errors from `errno.h` to _users_.

```
Fatal: unable to open config file: Stat: Access Denied.
Is there a repository at the following location?
s3:s3.amazonaws.com/backup.shrub.ca
```

Which config file was it trying to open? Was this a local or remote config file? What's the path to this file? Why is the error message asking the user questions? Why is the word "fatal" present? Why is the word "stat" present?

---

A good error message should:
* Clearly and accurately state the problem
* Give a reason for the error
* Suggest corrective action

Here are examples of more helpful errors:

>restic was unable to connect to the S3 bucket "backup.myserver.com," as no credentials were supplied. Please supply credentials, such as setting AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY or something something IAM and try again.

>restic was unable to connect to the S3 bucket "backup.myserver.com," as the credentials supplied were rejected by the server. Please ensure that AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY  (or IAM something something) are correct and try again.

**Comment by fd0** [comment]
You're right, the error message isn't great. I'd like to look into how the Upspin project [handles errors](https://commandcenter.blogspot.de/2017/12/error-handling-in-upspin.html), but that will require larger refactorings within restic.

**Comment by MarkOnDuty** [comment]
I just burned by this one. Wouldn't have scratched my head for so long if the error message was clearer. I definitely would have scratched my head for a lot longer if I hadn't found the discussion here.

**Comment by cdbessig** [comment]
@MarkOnDuty how did this help you?  I am scratching my head with this error with the same questions.  Did I miss the part where they actually dive into what the below means and how to fix it?

Fatal: unable to open config file: Stat: Access Denied.

**Comment by MarkOnDuty** [comment]
The last bit of @paulschreiber's initial comment was the key for me. The bit about the environment variables:
```
# export AWS_ACCESS_KEY_ID=xxxx
# export AWS_SECRET_ACCESS_KEY=xxxx
# export RESTIC_PASSWORD=xxxx
```
I just had the wrong value for the RESTIC_PASSWORD. Restic's error message sent me on a wild goose chase.

**Comment by sfxworks** [comment]
My case I had my access key in my AWS_SECRET_ACCESS_KEY variable. Google copypaste error to find relevant stack overflow or github discussions saves lives (by saving time :P) . 

Clarity would be cool.

**Comment by nbommu1** [comment]
this is going to be a big issue in my environment where we are going to switch everything to instance profile,  right now it is broken with version 0.9.5

```Fatal: unable to open config file: Stat: The specified key does not exist.
Is there a repository at the following location?
s3:s3.amazonaws.com/testbucket```

**Comment by nbommu1** [comment]
quick update, when I change s3 endpoint to

s3:s3.amazonaws.com/<BUCKET>/restic, I'm able take the backup using instance porfiles

**Comment by lgg42** [comment]
I also encountered this issue while trying to use instance roles. I was missing a permission in my IAM role policy: 

`s3:GetBucketLocation`

My restic version is 0.9.6 and my error was:

```
Fatal: unable to open config file: Stat: Head https://[bucket].s3.dualstack.us-east-1.amazonaws.com/[repo]/config: 301 response missing Location header
Is there a repository at the following location?
s3:s3.amazonaws.com/[bucket]/[repo]
```
After adding the missing permission it worked perfectly.

**Comment by VictorioBerra** [comment]
What IAM policies are needed on the bucket exactly? I am getting this error as well.

**Comment by MichaelEischer** [comment]
@VictorioBerra Which error exactly? Which restic version do you use? Do you use a region-specific URL or not?

**Comment by asheshambasta** [comment]
Got bitten by this one too, no progress so far.

**Comment by MichaelEischer** [comment]
Restic 0.17.0 will now fail with `no credentials found. Use '-o s3.unsafe-anonymous-auth=true' for anonymous authentication` if no credentials were specified. See https://github.com/restic/restic/issues/4707 for details.
